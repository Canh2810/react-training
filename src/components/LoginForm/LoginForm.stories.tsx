import { StoryObj, Meta } from '@storybook/react'

import LoginForm from '.'

export default {
  title: 'Components/LoginForm',
  component: LoginForm,
} as Meta

type Story = StoryObj<typeof LoginForm>

export const Default: Story = {}
