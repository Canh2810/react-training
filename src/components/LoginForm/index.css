.login-form {
  position: relative;
  width: 400px;
  text-align: center;
  display: flex;
  flex-direction: column;
  gap: 28px;
  background-color: var(--background-primary-color);
  padding: 40px;
  border-radius: 10px;
  box-shadow: 0px 12px 24px -6px rgba(24, 26, 42, 0.12);
}

.login-form__error-message {
  position: absolute;
  bottom: 10px;
  left: 120px;
}
