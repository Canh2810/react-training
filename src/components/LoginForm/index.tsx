// Libs
import { memo, useCallback } from 'react'
import { useForm, Controller, SubmitHandler } from 'react-hook-form'

// Components
import { Button, TextField, Typography } from '../index'

// Types
import {
  ButtonVariants,
  ILoginForm,
  LoginFormField,
  TextFieldTypes,
  TypoColors,
  TypoVariants,
} from '@types'

// Constants
import { ERROR_MESSAGES } from '@constants'

// Icons
import { Email, Lock } from '@assets/icons'

// Utils
import { checkEmail } from '@utils'

// Style
import './index.css'

export interface LoginFormProps {
  errorMessage?: string
  onSubmit: (data: ILoginForm) => void
}

const LoginForm = ({ errorMessage, onSubmit }: LoginFormProps) => {
  const loginFormField: LoginFormField[] = ['email', 'password']
  const { control, clearErrors, handleSubmit } = useForm<ILoginForm>({
    defaultValues: {
      email: '',
      password: '',
    },
  })

  const handleSubmitLoginForm: SubmitHandler<ILoginForm> = useCallback(
    (data: ILoginForm) => {
      onSubmit(data)
    },
    [onSubmit],
  )

  return (
    <form onSubmit={handleSubmit(handleSubmitLoginForm)} className="login-form">
      <Typography component="h1" variant={TypoVariants.HeadingMedium}>
        LOGIN
      </Typography>
      {loginFormField.map((item) => {
        const title = item === 'email' ? 'Email' : 'Password'
        return (
          <Controller
            key={item}
            control={control}
            name={item}
            rules={{
              required: ERROR_MESSAGES.FIELD_REQUIRED(title),
              ...(item === 'email' && {
                validate: (value: string) => checkEmail(value),
              }),
            }}
            render={({ field: { value, onChange }, fieldState: { error } }) => (
              <TextField
                icon={item === 'email' ? <Email /> : <Lock />}
                id={item}
                type={
                  item === 'email'
                    ? TextFieldTypes.Text
                    : TextFieldTypes.Password
                }
                name={item}
                placeholder={title}
                value={value}
                errorMessage={error?.message}
                onChange={(data) => {
                  clearErrors(item)
                  onChange(data)
                }}
              />
            )}
          />
        )
      })}
      <Button variant={ButtonVariants.Container} type="submit">
        Login
      </Button>
      {errorMessage && (
        <Typography
          className="login-form__error-message"
          color={TypoColors.Danger}
        >
          {errorMessage}
        </Typography>
      )}
    </form>
  )
}

export default memo(LoginForm)
