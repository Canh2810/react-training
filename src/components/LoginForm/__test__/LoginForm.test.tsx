import { fireEvent, render, waitFor, screen } from '@testing-library/react'
import LoginForm from '..'

describe('LoadingIndicator Component', () => {
  const mockOnSubmit = jest.fn()

  const setup = () => {
    const { getByPlaceholderText, getByText } = render(
      <LoginForm errorMessage="" onSubmit={mockOnSubmit} />,
    )

    const emailInput = getByPlaceholderText('Email')
    const passwordInput = getByPlaceholderText('Password')
    const submitButton = getByText('Login')

    return {
      emailInput,
      passwordInput,
      submitButton,
    }
  }
  it('should render correctly', () => {
    const { container } = render(<LoginForm onSubmit={mockOnSubmit} />)
    expect(container).toMatchSnapshot()
  })

  it('handles form submission with valid data', async () => {
    const { emailInput, passwordInput, submitButton } = setup()

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } })
    fireEvent.change(passwordInput, { target: { value: 'password123' } })
    fireEvent.click(submitButton)

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith({
        email: 'test@example.com',
        password: 'password123',
      })
    })
  })

  it('should display an error message for an invalid email', async () => {
    const { emailInput, submitButton } = setup()

    fireEvent.change(emailInput, { target: { value: 'invalidemail' } })
    fireEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Invalid format of email, please check.'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing email', async () => {
    const { passwordInput, submitButton } = setup()

    fireEvent.change(passwordInput, { target: { value: 'password@123' } })
    fireEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Email is required'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing password', async () => {
    const { emailInput, submitButton } = setup()

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } })
    fireEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Password is required'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing email and password', async () => {
    const { submitButton } = setup()
    fireEvent.click(submitButton)

    const emailErrorMessage = await waitFor(() =>
      screen.getByText('Email is required'),
    )

    const passwordErrorMessage = await waitFor(() =>
      screen.getByText('Password is required'),
    )

    expect(emailErrorMessage).toBeInTheDocument()
    expect(passwordErrorMessage).toBeInTheDocument()
  })
})
