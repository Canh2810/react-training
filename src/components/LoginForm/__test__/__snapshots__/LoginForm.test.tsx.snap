// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator Component should render correctly 1`] = `
<div>
  <form
    class="login-form"
  >
    <h1
      class=""
      style="font-family: work-sans; font-size: 20px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
    >
      LOGIN
    </h1>
    <div
      class="text-field "
      style="border: 1px solid; color: rgb(105, 106, 117); border-color: rgba(105, 106, 117, 0.30); background-color: rgb(255, 255, 255);"
    >
      <input
        id="email"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <svg
        fill="none"
        height="20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3.438 4.375a.937.937 0 0 0-.938.938v9.375c0 .517.42.937.938.937h13.124c.518 0 .938-.42.938-.938V5.313a.937.937 0 0 0-.938-.938H3.438Zm-2.188.938c0-1.209.98-2.188 2.188-2.188h13.124c1.209 0 2.188.98 2.188 2.188v9.375c0 1.208-.98 2.187-2.188 2.187H3.438a2.188 2.188 0 0 1-2.187-2.188V5.313Z"
          fill="#696A75"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M3.882 5.866a.625.625 0 0 1 .877-.11L10 9.834l5.241-4.076a.625.625 0 1 1 .768.986l-5.625 4.375a.625.625 0 0 1-.768 0L3.991 6.743a.625.625 0 0 1-.11-.877Z"
          fill="#696A75"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="text-field "
      style="border: 1px solid; color: rgb(105, 106, 117); border-color: rgba(105, 106, 117, 0.30); background-color: rgb(255, 255, 255);"
    >
      <input
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <svg
        class="ionicon"
        height="20px"
        viewBox="0 0 512 512"
        width="20px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M336 208v-95a80 80 0 0 0-160 0v95"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
        />
        <rect
          fill="none"
          height="272"
          rx="48"
          ry="48"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
          width="320"
          x="96"
          y="208"
        />
      </svg>
    </div>
    <button
      class="btn "
      style="background-color: rgb(75, 107, 251); color: rgb(255, 255, 255);"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
